[OpenMP] Add diagnostic for 'factor' width mismatch in 'unroll partial'

albus-droid opened the pull request:
This patch fixes #139268 by preventing an assertion failure when a user specifies a factor in `unroll partial` that either exceeds the signed 64-bit limit or the bit-width of the loop’s iteration variable.
 

  - Used `VerifyPositiveIntegerConstantInClause` to ensure the factor is strictly positive and fits in signed 64 bits. 
 
  -  Checking the factors bit-width to the iteration variable’s width, emitting a new `err_omp_unroll_factor_width_mismatch` diagnostic if it’s larger

  - Added some tests
  
  - Ran tests by calling `llvm-lit -v clang/test/OpenMP` 
  
![screen](https://github.com/user-attachments/assets/d9c08568-9973-48f2-918d-4ba037f1413c)

albus-droid committed:
Added two conditions to check for width of the factor variable

albus-droid committed:
Added diag error message and also formatted using clang-format

albus-droid committed:
Added two test cases

shafik commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14924,8 +14924,24 @@ StmtResult SemaOpenMP::ActOnOpenMPUnrollDirective(ArrayRef<OMPClause *> Clauses,
+      return StmtError();
+    }
+    // Checking if Itertor Variable Type can hold the Factor Width
+    if (FactorVal->getIntegerConstantExpr(Context)->getBitWidth() >
```
If we issue a diagnostic we will evaluate `FactorVal->getIntegerConstantExpr(Context)` twice. Let's just assign it to a local variable and use that in both places.

albus-droid replied:
I created two local variables `FactorValWidth` and `IteratorVWidth` to use in both places.

shafik commented on clang/test/OpenMP/unroll_messages.cpp:
```diff
@@ -58,8 +58,17 @@ void func(int n) {
+
+  // expected-error@+1 {{unroll factor has width 64 but the iteration variable 'int' is only 32 bits wide}}
+  #pragma omp unroll partial(0xFFFFFFFFF)
+  for (int i = 0; i < 10; i++)
```
What do we expect if `i` is *short* or *char* and the value does not fit? If this should fail we should test it.

albus-droid replied:
I realized I was using the wrong loop variable: the type still shows as i32 even after I changed it to char. I should be using OrigVar instead of IVTy. I’ll include that fix along with the other changes.

albus-droid replied:
I updated the code that determines the iterator variable’s width and type. Previously it didn’t adjust correctly for `long`, `short`, or `char`. Then I added new test cases and revised the existing ones. Now all of the tests pass.

albus-droid replied:
Also I have added new test to cover cases with short and char and revised the existing ones.

albus-droid committed:
Changed the Factor variable and Iterator Variable

albus-droid committed:
updated and added some new test cases

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14924,8 +14924,25 @@ StmtResult SemaOpenMP::ActOnOpenMPUnrollDirective(ArrayRef<OMPClause *> Clauses,
+    // Checking if Iterator Variable Type can hold the Factor Width
+    auto FactorValWidth = FactorVal->getIntegerConstantExpr(Context)->getActiveBits();
+    auto IteratorVWidth = Context.getTypeSize(OrigVar->getType());
+    if ( FactorValWidth > IteratorVWidth ) {
```
Formatting?

alexey-bataev reviewed (approved):
LG with a nit

shiltian commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14924,8 +14924,25 @@ StmtResult SemaOpenMP::ActOnOpenMPUnrollDirective(ArrayRef<OMPClause *> Clauses,
+    if (!VerifyPositiveIntegerConstantInClause(FactorVal, OMPC_partial,
+                                               /*StrictlyPositive=*/true,
+                                               /*SuppressExprDiags=*/false)
+             .isUsable()) {
+      return StmtError();
+    }
```
```suggestion
             .isUsable())
      return StmtError();
```

shiltian commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14924,8 +14924,25 @@ StmtResult SemaOpenMP::ActOnOpenMPUnrollDirective(ArrayRef<OMPClause *> Clauses,
+      return StmtError();
+    }
+    // Checking if Iterator Variable Type can hold the Factor Width
+    auto FactorValWidth = FactorVal->getIntegerConstantExpr(Context)->getActiveBits();
+    auto IteratorVWidth = Context.getTypeSize(OrigVar->getType());
```
no auto

shiltian commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14924,8 +14924,25 @@ StmtResult SemaOpenMP::ActOnOpenMPUnrollDirective(ArrayRef<OMPClause *> Clauses,
+
     Factor = FactorVal->getIntegerConstantExpr(Context)->getZExtValue();
     FactorLoc = FactorVal->getExprLoc();
+
```
unrelated changes

albus-droid committed:
Added the requested changes

albus-droid committed:
Added the requested formatting changes

albus-droid commented:
Thanks for the review. I’ve implemented all the requested changes but still have a few questions:

* Originally, using `IVTy` to query the iteration variable’s width always returned “32-bit int,” even if the loop was declared as `char` or `short`. As a result, errors only appeared when the factor exceeded 32 bits, and the diagnostic always said “int (32 bits).” However, when the loop variable was `long`, the code compiled successfully, so larger types were recognized. I’m not sure whether this was due to how I was reading the type or something else.

* In the current implementation I switched to using `OrigVar->getType()` to get the variable’s `QualType`. Now `getTypeSize()` correctly returns 8 bits for `char`, 16 for `short`, 64 for `long`, etc. That means code like

  ```cpp
  #pragma omp unroll partial(0xFFFFF)
  for (char i = 0; …) { }
  ```

  now rejects, whereas it used to compile.

* One remaining concern: the bit-width check treats all iteration variables as unsigned. It doesn’t account for signed types. Should I add a separate signed-range check, as suggested here in this [comment](https://github.com/llvm/llvm-project/issues/139268#issuecomment-2867068205)?

albus-droid commented:
Ping @AaronBallman — could you take a look at this when you have a moment? Thanks!

AaronBallman commented:
> Thanks for the review. I’ve implemented all the requested changes but still have a few questions:
> 
>     * Originally, using `IVTy` to query the iteration variable’s width always returned “32-bit int,” even if the loop was declared as `char` or `short`. As a result, errors only appeared when the factor exceeded 32 bits, and the diagnostic always said “int (32 bits).” However, when the loop variable was `long`, the code compiled successfully, so larger types were recognized. I’m not sure whether this was due to how I was reading the type or something else.
> 
>     * In the current implementation I switched to using `OrigVar->getType()` to get the variable’s `Type`. Now `getTypeSize()` correctly returns 8 bits for `char`, 16 for `short`, 64 for `long`, etc. That means code like
>       ```c++
>       #pragma omp unroll partial(0xFFFFF)
>       for (char i = 0; …) { }
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       now rejects, whereas it used to compile.
> 
>     * One remaining concern: the bit-width check treats all iteration variables as unsigned. It doesn’t account for signed types. Should I add a separate signed-range check, as suggested here in this [comment](https://github.com/llvm/llvm-project/issues/139268#issuecomment-2867068205)?

I think I need @alexey-bataev or @shiltian to weigh in on what the OpenMP specification expects in these cases. My naive take is that we should be rejecting that code which previously compiled and we should be accounting for signed types.

albus-droid committed:
Update SemaOpenMP.cpp

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

AaronBallman commented:
> Is something preventing this from being merged?

https://github.com/llvm/llvm-project/pull/139986#issuecomment-2900974777 is unanswered. CC @alexey-bataev @shiltian

albus-droid commented:
Thanks for the clarifications! I’ll try to get the changes in soon.
